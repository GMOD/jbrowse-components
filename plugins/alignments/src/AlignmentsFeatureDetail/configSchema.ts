import { ConfigurationSchema } from '@jbrowse/core/configuration'

export const configSchema = ConfigurationSchema('AlignmentsFeatureWidget', {})
