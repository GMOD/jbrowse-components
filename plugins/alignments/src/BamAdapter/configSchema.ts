import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'
import { types } from 'mobx-state-tree'

export default types.late(() =>
  ConfigurationSchema(
    'BamAdapter',
    {
      bamLocation: {
        type: 'fileLocation',
        defaultValue: { uri: '/path/to/my.bam' },
      },
      index: ConfigurationSchema('BamIndex', {
        indexType: {
          model: types.enumeration('IndexType', ['BAI', 'CSI']),
          type: 'stringEnum',
          defaultValue: 'BAI',
        },
        location: {
          type: 'fileLocation',
          defaultValue: { uri: '/path/to/my.bam.bai' },
        },
      }),
      chunkSizeLimit: {
        type: 'number',
        defaultValue: 100000000,
      },
      fetchSizeLimit: {
        type: 'number',
        defaultValue: 500000000,
      },
      sequenceAdapter: {
        type: 'frozen',
        defaultValue: null,
      },
    },
    { explicitlyTyped: true },
  ),
)
