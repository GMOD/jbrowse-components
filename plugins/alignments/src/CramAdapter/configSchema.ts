import PluginManager from '@jbrowse/core/PluginManager'
import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default (pluginManager: PluginManager) => {
  const { types } = pluginManager.lib['mobx-state-tree']
  return types.late(() =>
    ConfigurationSchema(
      'CramAdapter',
      {
        cramLocation: {
          type: 'fileLocation',
          defaultValue: { uri: '/path/to/my.cram' },
        },
        craiLocation: {
          type: 'fileLocation',
          defaultValue: { uri: '/path/to/my.cram.crai' },
        },
        sequenceAdapter: pluginManager.pluggableConfigSchemaType('adapter'),
      },
      { explicitlyTyped: true },
    ),
  )
}
