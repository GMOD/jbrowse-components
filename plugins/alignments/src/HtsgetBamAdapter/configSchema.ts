import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { types } from 'mobx-state-tree'

export default types.late(() =>
  ConfigurationSchema(
    'HtsgetBamAdapter',
    {
      htsgetBase: {
        type: 'string',
        defaultValue: '',
      },
      htsgetTrackId: {
        type: 'string',
        defaultValue: '',
      },
      sequenceAdapter: {
        type: 'frozen',
        defaultValue: null,
      },
    },
    { explicitlyTyped: true },
  ),
)
