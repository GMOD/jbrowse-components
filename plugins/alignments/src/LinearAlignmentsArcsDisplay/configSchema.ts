import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { linearBasicDisplayConfigSchemaFactory } from '@jbrowse/plugin-linear-genome-view'
import { types, Instance } from 'mobx-state-tree'
import PluginManager from '@jbrowse/core/PluginManager'

/**
 * #config LinearAlignmentsArcsDisplay
 */
function x() {} // eslint-disable-line @typescript-eslint/no-unused-vars

function configSchemaF(pluginManager: PluginManager) {
  // modify config schema to take in a sub coverage display
  return ConfigurationSchema(
    'LinearAlignmentsArcsDisplay',
    {
      /**
       * #slot
       */
      defaultRendering: {
        type: 'stringEnum',
        model: types.enumeration('Rendering', ['pileup']),
        defaultValue: 'pileup',
      },
      /**
       * #slot
       */
      renderers: ConfigurationSchema('RenderersConfiguration', {
        PileupRenderer:
          pluginManager.getRendererType('PileupRenderer').configSchema,
      }),
      /**
       * #slot
       */
      maxFeatureScreenDensity: {
        type: 'number',
        description: 'maximum features per pixel that is displayed in the view',
        defaultValue: 5,
      },

      /**
       * #slot
       */
      colorScheme: {
        type: 'stringEnum',
        model: types.enumeration('colorScheme', [
          'strand',
          'normal',
          'insertSize',
          'insertSizeAndOrientation',
          'mappingQuality',
          'tag',
        ]),
        description: 'color scheme to use',
        defaultValue: 'normal',
      },
    },
    {
      /**
       * #baseConfiguration
       */
      baseConfiguration: linearBasicDisplayConfigSchemaFactory(pluginManager),
      explicitlyTyped: true,
    },
  )
}

export type LinearAlignmentsArcsDisplayConfigModel = ReturnType<
  typeof configSchemaF
>
export type LinearAlignmentsArcsDisplayConfig =
  Instance<LinearAlignmentsArcsDisplayConfigModel>
export default configSchemaF
