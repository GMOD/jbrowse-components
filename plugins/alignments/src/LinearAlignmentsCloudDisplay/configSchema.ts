import PluginManager from '@jbrowse/core/PluginManager'
import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { linearBasicDisplayConfigSchemaFactory } from '@jbrowse/plugin-linear-genome-view'
import { types } from 'mobx-state-tree'

/**
 * #config LinearAlignmentsCloudDisplay
 */
function x() {} // eslint-disable-line @typescript-eslint/no-unused-vars

function configSchemaF(pluginManager: PluginManager) {
  return ConfigurationSchema(
    'LinearAlignmentsCloudDisplay',
    {
      /**
       * #slot
       */
      maxFeatureScreenDensity: {
        type: 'number',
        description: 'maximum features per pixel that is displayed in the view',
        defaultValue: 5,
      },

      /**
       * #slot
       */
      colorScheme: {
        type: 'stringEnum',
        model: types.enumeration('colorScheme', [
          'strand',
          'normal',
          'insertSize',
          'insertSizeAndOrientation',
          'mappingQuality',
          'tag',
        ]),
        description: 'color scheme to use',
        defaultValue: 'normal',
      },
    },
    {
      /**
       * #baseConfiguration
       */
      baseConfiguration: linearBasicDisplayConfigSchemaFactory(pluginManager),
      explicitlyTyped: true,
    },
  )
}

export default configSchemaF
