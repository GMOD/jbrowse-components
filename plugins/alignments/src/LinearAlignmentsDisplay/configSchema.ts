import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { baseLinearDisplayConfigSchema } from '@jbrowse/plugin-linear-genome-view'
import type PluginManager from '@jbrowse/core/PluginManager'

/**
 * #config LinearAlignmentsDisplay
 * has a "pileup" sub-display, where you can see individual reads and a
 * quantitative "snpcoverage" sub-display track showing SNP frequencies
 * extends
 * - [BaseLinearDisplay](../baselineardisplay)
 */
export default function configModelFactory(pluginManager: PluginManager) {
  return ConfigurationSchema(
    'LinearAlignmentsDisplay',
    {
      /**
       * #slot
       */
      pileupDisplay: pluginManager.getDisplayType('LinearPileupDisplay')!
        .configSchema,

      /**
       * #slot
       */
      snpCoverageDisplay: pluginManager.getDisplayType(
        'LinearSNPCoverageDisplay',
      )!.configSchema,

      /**
       * #slot
       */
      height: {
        type: 'number',
        defaultValue: 250,
      },
    },
    {
      /**
       * #baseConfiguration
       */
      baseConfiguration: baseLinearDisplayConfigSchema,
      explicitlyTyped: true,
    },
  )
}
