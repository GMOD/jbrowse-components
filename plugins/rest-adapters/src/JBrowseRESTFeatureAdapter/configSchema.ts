import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'JBrowseRESTFeatureAdapter',
  {
    location: {
      type: 'fileLocation',
      description: 'the base URL for the REST endpoints of this API',
      defaultValue: {
        uri: 'https://example.com/path/to/my/rest/endpoint',
        locationType: 'UriLocation',
      },
    },
    assemblyNames: {
      type: 'stringArray',
      description:
        'list of assembly names contained in this REST API. only used if the assembly_names optional resource is not implemented',
      defaultValue: [],
    },
    extraQuery: {
      description: 'additional URL query values to pass to the REST API',
      type: 'frozen',
      defaultValue: {},
    },
    optionalResources: ConfigurationSchema(
      'JBrowseRESTFeatureAdapterOptionalResources',
      {
        assembly_names: {
          description:
            'set to true if the REST endpoint provides a list of its configured assemblies',
          type: 'boolean',
          defaultValue: true,
        },
        has_data_for_reference: {
          description:
            'set to true if the REST endpoint provides a list of reference sequence names for its configured assemblies',
          type: 'boolean',
          defaultValue: true,
        },
        reference_sequences: {
          description:
            'set to true if the REST endpoint provides a list of reference sequence names for its configured assemblies',
          type: 'boolean',
          defaultValue: false,
        },
        region_stats: {
          description:
            'set to true if the REST endpoint provides region statistics',
          type: 'boolean',
          defaultValue: false,
        },
        region_feature_densities: {
          description:
            'set to true if the REST endpoint provides region feature densities',
          type: 'boolean',
          defaultValue: false,
        },
      },
    ),
  },
  { explicitlyTyped: true },
)
