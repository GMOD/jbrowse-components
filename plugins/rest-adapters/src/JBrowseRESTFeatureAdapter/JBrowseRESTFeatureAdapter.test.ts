import { toArray } from 'rxjs/operators'
import Adapter from './JBrowseRESTFeatureAdapter'
import configSchema from './configSchema'
import PluginManager from '@jbrowse/core/PluginManager'
import { firstValueFrom } from 'rxjs'
import {
  FeatureScoreStats,
  isFeatureCoverageStats,
  isFeatureScoreStats,
} from '@jbrowse/core/util/stats'

function mockFetchResponses(
  adapter: Adapter,
  mockResponses: Record<string, unknown>,
) {
  return (
    jest
      // @ts-ignore
      .spyOn(adapter, 'fetch')
      // @ts-ignore
      .mockImplementation(async (fetcher, url: string, signal) => {
        return {
          async json() {
            if (url in mockResponses) {
              return mockResponses[url]
            }
            throw new Error('no mock for ' + url)
          },
        }
      })
  )
}

function createTestAdapter(
  configuration: Record<string, unknown>,
  mockResponses: Record<string, unknown>,
) {
  const stubManager = new PluginManager()
  const adapter = new Adapter(
    configSchema.create(configuration),
    undefined,
    stubManager,
  )

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const fetchMock: any = mockFetchResponses(adapter, mockResponses)

  return { adapter, fetchMock }
}

const testFeatures1 = [{ start: 100, end: 200, refName: '21', uniqueId: 'foo' }]

test('adapter can fetch features and stats from mocked API with no region stats', async () => {
  const { adapter, fetchMock } = createTestAdapter(
    {
      location: { uri: 'https://mock.url/mock/url' },
      extraQuery: { extra_query: 42 },
      assemblyNames: ['volvox'],
      optionalResources: {
        assembly_names: false,
      },
    },
    {
      'https://mock.url/mock/url/features/volvox/21?start=34960388&end=35960388&extra_query=42':
        {
          features: testFeatures1,
        },
      'https://mock.url/mock/url/volvox/reference_sequences?extra_query=42': [
        '21',
      ],
      'https://mock.url/mock/url/has_data_for_reference/volvox/ctgA?extra_query=42':
        false,
      'https://mock.url/mock/url/has_data_for_reference/volvox/21?extra_query=42':
        true,
    },
  )

  const features = adapter.getFeatures({
    assemblyName: 'volvox',
    refName: '21',
    start: 34960388,
    end: 35960388,
  })

  const featuresArray = await firstValueFrom(features.pipe(toArray()))
  expect(featuresArray.length).toBe(testFeatures1.length)
  expect(featuresArray[0].get('refName')).toBe('21')
  expect(featuresArray[0].id()).toBe('foo')
  // @ts-ignore
  expect(fetchMock.mock.calls[0][1]).toBe(
    'https://mock.url/mock/url/features/volvox/21?start=34960388&end=35960388&extra_query=42',
  )
  expect(await adapter.getAssemblyNames()).toStrictEqual(['volvox'])
  expect(
    await adapter.hasDataForRefName('ctgA', { assemblyName: 'volvox' }),
  ).toBe(false)
  expect(
    await adapter.hasDataForRefName('21', { assemblyName: 'volvox' }),
  ).toBe(true)
  expect(
    await adapter.hasDataForRefName('21', { assemblyName: 'volvox' }),
  ).toBe(true)
  expect(await adapter.hasDataForRefName('21')).toBe(true)

  expect(
    await adapter.getRegionStats({
      assemblyName: 'volvox',
      refName: '21',
      start: 34960388,
      end: 35960388,
    }),
  ).toMatchSnapshot()
})

test('adapter can fetch features and stats from mocked API with region stats and REST assembly names', async () => {
  const { adapter } = createTestAdapter(
    {
      location: { uri: 'https://mock.url/mock/url' },
      optionalResources: {
        assembly_names: true,
        region_stats: true,
        has_data_for_reference: false,
        reference_sequences: true,
      },
    },
    {
      'https://mock.url/mock/url/features/volvox/21?start=34960388&end=35960388':
        {
          features: [],
        },
      'https://mock.url/mock/url/assembly_names': ['zonker'],
      'https://mock.url/mock/url/reference_sequences/volvox': [],
      'https://mock.url/mock/url/reference_sequences/zonker': ['21'],
      'https://mock.url/mock/url/has_data_for_reference/volvox/ctgA': false,
      'https://mock.url/mock/url/has_data_for_reference/volvox/21': false,
      'https://mock.url/mock/url/has_data_for_reference/zonker/21': true,
      'https://mock.url/mock/url/stats/region/zonker/2?start=1&end=20': {
        featureCount: 20,
        basesCovered: 400,
      },
      'https://mock.url/mock/url/stats/region/zonker/3?start=400&end=500': {
        featureCount: 20,
        basesCovered: 100,
        scoreSum: 400,
        scoreSumSquares: 10000,
        scoreMax: 100,
        scoreMin: 42,
      },
    },
  )

  const features = adapter.getFeatures({
    assemblyName: 'volvox',
    refName: '21',
    start: 34960388,
    end: 35960388,
  })

  const featuresArray = await firstValueFrom(features.pipe(toArray()))
  expect(featuresArray.length).toBe(0)

  expect(await adapter.getAssemblyNames()).toStrictEqual(['zonker'])
  expect(await adapter.getRefNames({ assemblyName: 'volvox' })).toStrictEqual(
    [],
  )

  expect(
    await adapter.hasDataForRefName('ctgA', { assemblyName: 'volvox' }),
  ).toBe(false)
  expect(
    await adapter.hasDataForRefName('21', { assemblyName: 'volvox' }),
  ).toBe(false)
  expect(
    await adapter.hasDataForRefName('21', { assemblyName: 'zonker' }),
  ).toBe(true)

  // stats with no scores
  const stats1 = await adapter.getRegionStats({
    assemblyName: 'zonker',
    refName: '2',
    start: 1,
    end: 20,
  })
  expect(isFeatureCoverageStats(stats1)).toBeTruthy()
  expect(isFeatureScoreStats(stats1)).toBeFalsy()
  expect(stats1).toMatchSnapshot()

  // stats with scores
  const stats2 = (await adapter.getRegionStats({
    assemblyName: 'zonker',
    refName: '3',
    start: 400,
    end: 500,
  })) as FeatureScoreStats
  expect(isFeatureCoverageStats(stats2)).toBeTruthy()
  expect(isFeatureScoreStats(stats2)).toBeTruthy()
  expect(stats2.featureDensity).toBeDefined()
  expect(stats2.scoreMean).toBeDefined()
  expect(stats2).toMatchSnapshot()
})
