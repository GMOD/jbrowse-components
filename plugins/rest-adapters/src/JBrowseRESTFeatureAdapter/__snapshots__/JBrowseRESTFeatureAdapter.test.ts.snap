// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features and stats from mocked API with no region stats 1`] = `
{
  "basesCovered": 1000000,
  "featureCount": 1,
  "featureDensity": 0.000001,
}
`;

exports[`adapter can fetch features and stats from mocked API with region stats and REST assembly names 1`] = `
{
  "basesCovered": 400,
  "featureCount": 20,
  "featureDensity": 0.05,
}
`;

exports[`adapter can fetch features and stats from mocked API with region stats and REST assembly names 2`] = `
{
  "basesCovered": 100,
  "featureCount": 20,
  "featureDensity": 0.2,
  "scoreMax": 100,
  "scoreMean": 20,
  "scoreMin": 42,
  "scoreStdDev": 10.25978352085154,
  "scoreSum": 400,
  "scoreSumSquares": 10000,
}
`;
